packages/datetime/src/dateRangeSelectionStrategy.ts

```diff
@@ -0,0 +1,155 @@
+
+import {
+    areSameDay,
+    DateRange,
+    DateRangeBoundary,
+} from "./common/dateUtils";
+
+export interface IDateRangeSelectionState {
+    /**
+     * The boundary that would be modified by clicking the provided `day`.
+     * May be different from `boundaryToModify` in some special cases
+     * (e.g. when hovering over the other boundary's selected date).
+     */
+    boundary?: DateRangeBoundary;
+
+    /**
+     * The date range that would be selected after clicking the provided `day`.
+     */
+    dateRange: DateRange;
+};
+
+export class DateRangeSelectionStrategy {
+    public static getNextState(currentRange: DateRange,
+                               day: Date,
+                               boundaryToModify?: DateRangeBoundary,
+                               allowSingleDayRange?: boolean) {
+        if (boundaryToModify != null) {
+            return this.getNextStateForBoundary(currentRange, day, boundaryToModify, allowSingleDayRange);
+        } else {
+            return this.getDefaultNextState(currentRange, day, allowSingleDayRange);
+        }
+    }
+
+    private static getNextStateForBoundary(currentRange: DateRange,
+                                           day: Date,
+                                           boundaryToModify: DateRangeBoundary,
+                                           allowSingleDayRange: boolean) {
+        const boundary = boundaryToModify;
```

Comment:
Because then you have this beautiful parallel structure of variable names within the function:

`boundary` and `otherBoundary`
`boundaryDate` and `otherBoundaryDate`

...but `boundaryToModify` is used for the parameter, because it more clearly communicates what the developer should pass in. This is worth a comment.